File: app/src/main/java/com/soya/launcher/ui/dialog/TimePickerDialog.java
package com.soya.launcher.ui.dialog;

import android.os.Bundle;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.leanback.widget.HorizontalGridView;

import com.soya.launcher.R;
import com.soya.launcher.adapter.DateSelectAdapter;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

public class TimePickerDialog extends SingleDialogFragment implements View.OnClickListener {
    public static final String TAG = "TimePickerDialog";
    public static TimePickerDialog newInstance() {

        Bundle args = new Bundle();

        TimePickerDialog fragment = new TimePickerDialog();
        fragment.setArguments(args);
        return fragment;
    }

    private final Calendar calendar = Calendar.getInstance();

    private HorizontalGridView mHourGrid;
    private HorizontalGridView mMinthGrid;
    private HorizontalGridView mSencondGrid;
    private View mCancelView;
    private View mConfirmView;
    private ImageView mBlur;
    private View mRootView;

    private DateSelectAdapter mHourAdapter;
    private DateSelectAdapter mMinuteAdapter;
    private DateSelectAdapter mSecondAdapter;

    private Callback callback;

    @Override
    protected int getLayout() {
        return R.layout.dialog_time_picker;
    }

    @Override
    protected void init(LayoutInflater inflater, View view) {
        super.init(inflater, view);
        mHourGrid = view.findViewById(R.id.hour);
        mMinthGrid = view.findViewById(R.id.minute);
        mSencondGrid = view.findViewById(R.id.second);
        mCancelView = view.findViewById(R.id.cancel);
        mConfirmView = view.findViewById(R.id.confirm);
        mBlur = view.findViewById(R.id.blur);
        mRootView = view.findViewById(R.id.root);

        List<Integer> years = new ArrayList<>();
        for (int i = 0; i < 24; i++){
            years.add(i);
        }
        List<Integer> months = new ArrayList<>();
        for (int i = 0; i <= 59; i++){
            months.add(i);
        }
        List<Integer> sencods = new ArrayList<>();
        for (int i = 0; i <= 59; i++){
            sencods.add(i);
        }
        mHourAdapter = new DateSelectAdapter(getActivity(), inflater, years, newHourCallback());
        mMinuteAdapter = new DateSelectAdapter(getActivity(), inflater, months, newMinthCallback());
        mSecondAdapter = new DateSelectAdapter(getActivity(), inflater, sencods, newSecondCallback());
    }

    @Override
    protected void initBefore(LayoutInflater inflater, View view) {
        super.initBefore(inflater, view);
        mCancelView.setOnClickListener(this);
        mConfirmView.setOnClickListener(this);
    }

    @Override
    protected void initBind(LayoutInflater inflater, View view) {
        super.initBind(inflater, view);
        int hourIndex = calendar.get(Calendar.HOUR_OF_DAY);
        int minthIndex = calendar.get(Calendar.MINUTE);
        int secondIndex = calendar.get(Calendar.SECOND);

        mHourGrid.setAdapter(mHourAdapter);
        mMinthGrid.setAdapter(mMinuteAdapter);
        mSencondGrid.setAdapter(mSecondAdapter);

        mHourGrid.setRowHeight(ViewGroup.LayoutParams.WRAP_CONTENT);
        mMinthGrid.setRowHeight(ViewGroup.LayoutParams.WRAP_CONTENT);
        mSencondGrid.setRowHeight(ViewGroup.LayoutParams.WRAP_CONTENT);

        mHourGrid.setSelectedPosition(hourIndex);
        mMinthGrid.setSelectedPosition(minthIndex);
        mSencondGrid.setSelectedPosition(secondIndex);

        mHourAdapter.setSelect(mHourAdapter.getDataList().get(mHourGrid.getSelectedPosition()));
        mMinuteAdapter.setSelect(mMinuteAdapter.getDataList().get(mMinthGrid.getSelectedPosition()));
        mSecondAdapter.setSelect(mSecondAdapter.getDataList().get(mSencondGrid.getSelectedPosition()));
    }

    @Override
    protected int getGravity() {
        return Gravity.CENTER;
    }

    @Override
    public boolean isMaterial() {
        return false;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        //blur(mRootView, mBlur);
    }

    @Override
    protected int[] getWidthAndHeight() {
        return new int[]{ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT};
    }

    @Override
    protected float getDimAmount() {
        return 0;
    }

    private DateSelectAdapter.Callback newHourCallback(){
        return new DateSelectAdapter.Callback() {
            @Override
            public void onClick(Integer bean) {
                calendar.set(Calendar.HOUR_OF_DAY, bean);
            }
        };
    }

    private DateSelectAdapter.Callback newMinthCallback(){
        return new DateSelectAdapter.Callback() {
            @Override
            public void onClick(Integer bean) {
                calendar.set(Calendar.MINUTE, bean - 1);
            }
        };
    }

    private DateSelectAdapter.Callback newSecondCallback(){
        return new DateSelectAdapter.Callback() {
            @Override
            public void onClick(Integer bean) {
                calendar.set(Calendar.SECOND, bean - 1);
            }
        };
    }

    @Override
    public void onClick(View v) {
        if (v.equals(mCancelView)){
            dismiss();
        }else if (v.equals(mConfirmView)){
            if (callback != null) callback.onConfirm(calendar.getTimeInMillis());
            dismiss();
        }
    }

    public void setCallback(Callback callback) {
        this.callback = callback;
    }

    public interface Callback{
        void onConfirm(long timeMills);
    }
}

File: app/src/main/java/com/soya/launcher/ui/dialog/TimePickerDialog.kt
package com.soya.launcher.ui.dialog

import android.os.Bundle
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import com.drake.brv.utils.mutable
import com.drake.brv.utils.setup
import com.shudong.lib_base.ext.clickNoRepeat
import com.soya.launcher.R
import com.soya.launcher.databinding.DialogTimePickerBinding
import com.soya.launcher.databinding.HolderDateSelectBinding
import java.util.Calendar

class TimePickerDialog : SingleDialogFragment<DialogTimePickerBinding>(), View.OnClickListener {
    private val calendar: Calendar = Calendar.getInstance()

   


    private var callback: Callback? = null


    private var selectIndexHour: Int? = null
    private var selectIndexMinute: Int? = null
    private var selectIndexSecond: Int? = null
    override fun init( view: View) {

        val hours: MutableList<Int> = ArrayList()
        for (i in 0..23) {
            hours.add(i)
        }
        val minutes: MutableList<Int> = ArrayList()
        for (i in 0..59) {
            minutes.add(i)
        }
        val sencods: MutableList<Int> = ArrayList()
        for (i in 0..59) {
            sencods.add(i)
        }

        binding.hour.setup {
            addType<Int>(R.layout.holder_date_select)
            onBind {
                val dto = getModel<Int>()
                val binding = getBinding<HolderDateSelectBinding>()
                binding.title.isSelected = dto == selectIndexHour
                binding.title.text = dto.toString()
                itemView.clickNoRepeat {
                    selectIndexHour = dto
                    notifyItemRangeChanged(0, mutable.size)
                    calendar[Calendar.HOUR_OF_DAY] = dto
                }
            }
        }.models = hours

        binding.minute.setup {
            addType<Int>(R.layout.holder_date_select)
            onBind {
                val dto = getModel<Int>()
                val binding = getBinding<HolderDateSelectBinding>()
                binding.title.isSelected = dto == selectIndexMinute
                binding.title.text = dto.toString()
                itemView.clickNoRepeat {
                    selectIndexMinute = dto
                    notifyItemRangeChanged(0, mutable.size)
                    calendar[Calendar.MINUTE] = dto - 1
                }
            }
        }.models = minutes

        binding.second.setup {
            addType<Int>(R.layout.holder_date_select)
            onBind {
                val dto = getModel<Int>()
                val binding = getBinding<HolderDateSelectBinding>()
                binding.title.isSelected = dto == selectIndexSecond
                binding.title.text = dto.toString()
                itemView.clickNoRepeat {
                    selectIndexSecond = dto
                    notifyItemRangeChanged(0, mutable.size)
                    calendar[Calendar.SECOND] = dto - 1
                }
            }
        }.models = sencods

    }

    override fun initBefore(view: View) {
        binding.cancel.setOnClickListener(this)
        binding.confirm.setOnClickListener(this)
    }

    override fun initBind( view: View) {
        val hourIndex = calendar[Calendar.HOUR_OF_DAY]
        val minthIndex = calendar[Calendar.MINUTE]
        val secondIndex = calendar[Calendar.SECOND]


        binding.hour.setRowHeight(ViewGroup.LayoutParams.WRAP_CONTENT)
        binding.minute.setRowHeight(ViewGroup.LayoutParams.WRAP_CONTENT)
        binding.second.setRowHeight(ViewGroup.LayoutParams.WRAP_CONTENT)

        binding.hour.selectedPosition = hourIndex
        binding.minute.selectedPosition = minthIndex
        binding.second.selectedPosition = secondIndex

        selectIndexHour = binding.hour.mutable[binding.hour.selectedPosition] as Int?
        selectIndexMinute = binding.minute.mutable[binding.hour.selectedPosition] as Int?
        selectIndexSecond = binding.second.mutable[binding.second.selectedPosition] as Int?

    }

    override fun getGravity(): Int {
        return Gravity.CENTER
    }

    override fun isMaterial(): Boolean {
        return false
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        blur(binding.root, binding.blur)
    }

    override fun getWidthAndHeight(): Pair<Int,Int> {
        return ViewGroup.LayoutParams.MATCH_PARENT to ViewGroup.LayoutParams.MATCH_PARENT
    }

    override fun getDimAmount(): Float {
        return 0f
    }


    override fun onClick(v: View) {
        if (v == binding.cancel) {
            dismiss()
        } else if (v == binding.confirm) {
            if (callback != null) callback!!.onConfirm(calendar.timeInMillis)
            dismiss()
        }
    }

    fun setCallback(callback: Callback?) {
        this.callback = callback
    }

    interface Callback {
        fun onConfirm(timeMills: Long)
    }

    companion object {
        const val TAG: String = "TimePickerDialog"
        fun newInstance(): TimePickerDialog {
            val args = Bundle()

            val fragment = TimePickerDialog()
            fragment.arguments = args
            return fragment
        }
    }
}
